benchmarks/bench_ufunc.py

```diff
@@ -1,8 +1,45 @@
+import functools
+import os
+from typing import List, Mapping, Optional, Tuple
+
 import numpy as np
 import pnumpy
 pnumpy.initialize()
 
 
+########
+# TODO: Additional benchmarking aspects that should be incorporated into the benchmarks below:
+#       * start-of-array address alignment to 1/2/4/8/16/32/64/128-byte boundary to look for perf issues
+#         caused by things like missing some loop-peeling step (to process the first part of an array to
+#         get to an SIMD-vector-aligned boundary).
+#       * striding -- e.g. compare the performance of some function operating on an array of N elements
+#         vs. an array with the same number of elements but using a non-default stride (e.g. create an array
+#         2/3/4x the size then use slicing syntax with an explicit step value); maybe also include a 3rd
+#         comparison point where we use slicing syntax on the original array to take every other element
+#         (since that array/view would cover the same amount of memory but a smaller number of elements).
+#       * Can we report additional information at the process level (of the asv invocation)? Or would this
+#         need to be implemented in asv itself? It would be nice to use 'hwloc' to capture system topology
+#         info and save it as e.g. some metadata of the benchmark results, so we can consider doing some
+#         post-processing of the benchmark results to normalize by things like L2/L3 cache size, number of cores,
+#         SMT on/off (and SMT level), number of memory channels, etc.
```

Comment:
There [is a way](https://asv.readthedocs.io/en/stable/using.html#machine-information) to annotate the "machine" you define when you run asv. But this airspeed-velocity/asv#944 may be relvant.

Reply:
I was thinking of something more along the lines of airspeed-velocity/asv#791. I think your comment in that issue about saving the supported features (e.g. captured from CPUID flags) of the CPU used to run the numpy benchmarks is along the same lines -- I'd want to save that information along with the CPU/cache/memory topology of the machine to be able to get a better idea of how well (or not) the benchmarked functions are utilizing the hardware. (There are other ways to make that determination, e.g. by looking at CPU hardware counters, but this will be more portable across CPU architectures.)